src/disposer/AsyncDisposerSet.ts

```diff
@@ -0,0 +1,26 @@
+import { AsyncDisposer } from './AsyncDisposer';
+import { IAsyncDisposable } from './types';
+
+export class AsyncDisposerSet implements IAsyncDisposable {
+  protected readonly disposables: Set<IAsyncDisposable> = new Set();
+
+  get disposed(): boolean {
+    return this.disposables.size === 0;
+  }
+
+  public async dispose(): Promise<void> {
+    while (!this.disposed) {
```

Comment:
yes because someone could push to the `AsyncDisposerSet` inside one of the `dispose()` commands

Reply:
> yes because someone could push to the AsyncDisposerSet inside one of the dispose() commands

this may end up in infinite loop, you sure we really want this?
I'd just throw if the code inside of `dispose` method modifies the collection.

Reply:
Yes, it's a risk we have to take. Because there's also a (much more likely possibility) of an external event adding to the set during the dispose phase.

Reply:
Fair enough. I'd still throw, but I'm okay with trying to loop until every disposable is disposed.